packages/lime-proto-babeld/files/usr/lib/lua/lime/proto/babeld.lua

```diff
@@ -103,21 +114,70 @@ function babeld.setup_interface(ifname, args)
 	local vlanProto = args[3] or "8021ad"
 	local nameSuffix = args[4] or "_babeld"
 
+
+	--! If Babeld is without VLAN (vlanId is 0) it should run directly on plain
+	--! ethernet interfaces, but the ones which are inside of the LAN bridge
+	--! (e.g. eth0 or eth0.1) cannot have an IPv6 Link-Local and Babeld needs it.
+	--! So Babeld has to run on the bridge interface br-lan
+	local isIntoLAN = false
+	local addIPtoIf = true
+	for _,v in pairs(args["deviceProtos"]) do
+		if v == "lan" then
+			isIntoLAN = true
+		--! would be weird to add a static IP to the WAN interface
+		elseif v == "wan" then
+			addIPtoIf = false
+		end
+	end
+
+	if ifname:match("^wlan") then
+		--! currently (2019-10-12) mode-ap and mode-apname have an hardcoded
+		--! "option network lan" so they are always in the br-lan bridge
+		if ifname:match("^wlan.*ap$") or ifname:match("^wlan.*apname$") then
+			isIntoLAN = true
+
+		--! all the WLAN interfaces are ignored by proto-lan
+		--! so they are not in the bridge even if proto-lan is present
+		--! (except mode-ap and mode-apname as mentioned above)
+		else
+			isIntoLAN = false
+		end
+	end
+
+	if tonumber(vlanId) == 0 and isIntoLAN then
+		utils.log("Rather than "..ifname..
+		  ", adding br-lan into Babeld interfaces")
+		ifname = "br-lan"
+		--! br-lan has already an IPv4, no need to add it
+		addIPtoIf = false
+	end
+
 	local owrtInterfaceName, linuxVlanIfName, owrtDeviceName =
 	  network.createVlanIface(ifname, vlanId, nameSuffix, vlanProto)
 
-	local ipv4, _ = network.primary_address()
-
 	local uci = config.get_uci_cursor()
 
-	if(vlanId ~= 0 and ifname:match("^eth")) then
+	if tonumber(vlanId) ~= 0 and ifname:match("^eth") then
```

Comment:
Hmmm.... can vlanId be a string?

Reply:
It is a very bad idea to set it as a string, but if an user sets it as a string, network.lua at this line

https://github.com/libremesh/lime-packages/blob/d81c12cdf56f0370600999adf586265c7cd7668b/packages/lime-system/files/usr/lib/lua/lime/network.lua#L290

passes the string to the proto, and no protocol force it to be a number. e.g. babeld just takes it

https://github.com/libremesh/lime-packages/blob/d81c12cdf56f0370600999adf586265c7cd7668b/packages/lime-proto-babeld/files/usr/lib/lua/lime/proto/babeld.lua#L102

lime-config returns no error if a string is used and results in a nil value here

https://github.com/libremesh/lime-packages/blob/d81c12cdf56f0370600999adf586265c7cd7668b/packages/lime-system/files/usr/lib/lua/lime/network.lua#L339

and interfaces with appended a "_nil" string here

https://github.com/libremesh/lime-packages/blob/d81c12cdf56f0370600999adf586265c7cd7668b/packages/lime-system/files/usr/lib/lua/lime/network.lua#L363

I'm going to prepare a PR about this :)

Reply:
Not only this, it is always a string now as how it is parsed from the config. So yes, the conversion to numeric is always needed... (the problem is that we dont have a single test for this for example with vlanid=0)